File: 200-rotateMatrixWithClockWise.py

def transpose(A):
    if not A:
        return None

    R,C=len(A),len(A[0])
    
    T=[[0 for _ in range(R)] for _ in range(C)]
    
    for r in range(R):
        for c in range(C):
            T[c][r]=A[r][c]
    
    return T

def rotate(M):
    M[:]=transpose(M)
    
    r=len(M)
    for i in range(r):
        M[i]=M[i][::-1]
    return M

A=[[1,2,3],[4,5,6],[7,8,9]]
R=rotate(A)
print(R)

File: 200-rotateMatrixWithClockWise.cpp
//rotate one matrix as 45 degree

#include<vector>
#include<iostream>

using DT=std::vector<std::vector<int>>;




DT transpose(DT& M)
{
    int R=M.size(),C=M[0].size();
    DT N(C,std::vector<int>(R,0));
    for(int r=0;r<R;++r)
    {
        for(int c=0;c<C;++c)
        {
            N[c][r]=M[r][c];
        }
    }
    return N;
}


DT rotateMatrix45Degree(DT& M)
{
    DT N=transpose(M);
    int R=N.size(),C=N[0].size();
    //swap each line
    for(int r=0;r<R;++r)
    {
        int s=0,end=N[r].size()-1;
        while(s<end)
        {
            int tmp=N[r][s];
            N[r][s]=N[r][end];
            N[r][end]=tmp;
            s+=1;
            end-=1;
        }
    }
    return N;
}

int main()
{
    DT M{{1,2,3},{4,5,6},{7,8,9}};
    for(auto r:M)
    {
        std::cout<<'\n';
        for(auto v:r)
            std::cout<<v<<',';
    }
    std::cout<<"\n-------------\n";
    DT N=rotateMatrix45Degree(M);
        for(auto r:N)
    {
        std::cout<<'\n';
        for(auto v:r)
            std::cout<<v<<',';
    }
}